sync blogs 2024-07-01-11-03-13 (#1313)
emqx/blog: en/202407/exploring-geo-distribution-in-emqx-for-enhanced-scalability.md

## A Cluster 12600 km Wide

As you might already know, EMQX 5 follows a new deployment model, where the cluster consists of two types of nodes: *core* and *replicant*. This model was specifically designed to support a broader range of deployment scenarios, including geo-distribution, though with some pitfalls, which we will explore later. Under this model, a cluster consists of 2 types of nodes: core nodes that manage essential parts of the shared state and replicant nodes that do not participate in the state management but simply replicate state changes made by the core nodes.

Nevertheless, let us first deploy a cluster in a more traditional fashion, so we can better understand where the benefits of the new model could come from. Our “traditional” cluster will consist of 3 core nodes geographically distributed over 3 continents.

| **Location**             | **Node name**        |
| :----------------------- | :------------------- |
| Frankfurt (eu-central-1) | `emqx@euc1.emqx.dev` |
| Hong Kong (ap-east-1)    | `emqx@ape1.emqx.dev` |
| Cape Town (af-south-1)   | `emqx@afs1.emqx.dev` |

> ![EMQX Geo-Distribution](https://assets.emqx.com/images/460cdbd66a7e51e84f988f1574874025.png)

Reviewer: This map actually illustrates the cluster that is subject of "Replicants to the Rescue" section, I believe it should be moved here. This deployment should have "triangular" shape, basically joining 3 geo-points mentioned above.

We were lucky here: if there were any connectivity troubles, they were mild enough to be dealt with by the TCP stack. If the network had exhibited more severe instability to cause one of the nodes to become unreachable, transactions would have started to time out. Transactions of this nature could have grabbed locks on the records, which would have remained held until the transaction timed out. This scenario would have effectively blocked all the other transactions involving conflicting records, thus stalling the subscriptions altogether.

## Avoiding the Conflicts

> Luckily, this *v1* schema we've manually enabled for our cluster is no longer the default, and the new [*v2*](https://github.com/emqx/emqx/pull/11524) schema was considered pretty robust and performant enough to take its place as the default starting from [EMQX 5.4.0](https://github.com/emqx/emqx/releases/tag/v5.4.0). The main difference that radically improves the situation is that it's *conflict-free*: no matter how messy and complex are topic filters your clients are using, there's zero chance managing them will involve write conflicts. You don't need to worry about potential clashes in subscription patterns anymore.

Reviewer: ```suggestion
Luckily, this *v1* schema we've manually enabled for our cluster is no longer the default, and the new [*v2*](https://github.com/emqx/emqx/pull/11524) schema was considered pretty robust and performant enough to take its place as the default starting from [EMQX 5.4.0](https://github.com/emqx/emqx/releases/tag/v5.4.0). The main difference that radically improves the situation is that it's *conflict-free*: no matter how messy and complex topic filters your clients are using, there's zero chance managing them will involve write conflicts. You don't need to worry about potential clashes in subscription patterns anymore.
```

```shell
$ mqttx bench sub -h ape1.emqx.dev --count 100 --interval 0 --topic t/mqttx/+/%i/sub/+/#
[11:04:35 AM] › | Start the subscribe benchmarking, connections: 100, req interval: 0ms, topic: t/mqttx/+/%i/sub/+/#
✔  success   [100/100] - Subscribed to t/mqttx/+/100/sub/+/#
[11:04:37 AM] › | Created 100 connections in 1.091s
```

That's a notable improvement. Remember that we started with 100 clients taking more than 100 seconds to do the same thing.

## Hiding the Latency

Before, we briefly mentioned that the throughput may be actually artificially limited by the default number of processes in a pool acting as serialization points to avoid race conditions. Increasing the load 10 times should help us observe the effects of this limit.

```shell
$ mqttx bench sub -h ape1.emqx.dev --count 1000 --interval 0 --topic t/mqttx/+/%i/sub/+/#
[11:42:39 PM] › ℹ  Start the subscribe benchmarking, connections: 1000, req interval: 0ms, topic: t/mqttx/+/%i/sub/+/#
✔  success   [1000/1000] - Subscribed to t/mqttx/+/1000/sub/+/#
[11:42:43 PM] › ℹ  Created 1000 connections in 4.329s
```

> With 10 times more network traffic it's likely that the misbehaving network is mostly to blame, but the total time is still 4 times worse. Can we do better? Thankfully, there's another improvement we had been working on and that landed in [EMQX 5.5.0](https://github.com/emqx/emqx/releases/tag/v5.5.0) release: [batch synchronization](https://github.com/emqx/emqx/pull/12329) of routing table updates. It is designed to better utilize the available network throughput when the communication is reliable, without increasing latency of operations when it's already too high.

Reviewer: ```suggestion
With 10 times more network traffic it's likely that the misbehaving network is mostly to blame, but the total time is still 4 times worse. Can we do better? Thankfully, there's another improvement we have been working on and that landed in the [EMQX 5.5.0](https://github.com/emqx/emqx/releases/tag/v5.5.0) release: [batch synchronization](https://github.com/emqx/emqx/pull/12329) of routing table updates. It is designed to better utilize the available network throughput when the communication is reliable, without increasing latency of operations when it's already too high.
```

This feature is not yet enabled by default, but it's pretty easy to turn on.

```
broker.routing.batch_sync.enable_on = replicant
```

This snippet enables batch synchronization only on replicants. Setting it to `all` would enable it on all nodes, and in general will likely be beneficial for a wide variety of workloads, mostly thanks to the broker pool being freed up from doing extra synchronization work. In our case, the effects would not be as pronounced and harder to illustrate.

> The default is `none`, and the reason for that is safety. We want to provide smooth rolling upgrade experience, and enabling it *before* rolling out EMQX 5.7.0 over, let's say, 5.4.1 could lead to some temporary unavailability. If you find yourself upgrading from EMQX 5.5.0 or later however, you should not worry about enabling it right away.

Reviewer: ```suggestion
The default is `none`, and the reason for that is safety. We want to provide a smooth rolling upgrade experience, and enabling it *before* rolling out EMQX 5.7.0 over, let's say, 5.4.1 could lead to some temporary unavailability. If you find yourself upgrading from EMQX 5.5.0 or later, you should not worry about enabling it right away.
```

Ok then, let's see how it goes now.

```shell
$ mqttx bench sub -h ape1.emqx.dev --count 1000 --interval 0 --topic t/mqttx/+/%i/sub/+/#
[11:46:21 PM] › ℹ  Start the subscribe benchmarking, connections: 1000, req interval: 0ms, topic: t/mqttx/+/%i/sub/+/#
✔  success   [1000/1000] - Subscribed to t/mqttx/+/1000/sub/+/#
[11:46:24 PM] › ℹ  Created 1000 connections in 2.585s
```

Pretty significant improvement. The network throughput is now used more efficiently, which in turn translates to better `SUBSCRIBE` throughput. Roughly the same effect could most likely be achieved by inflating process pool at the cost of more bookkeeping and increased memory usage, but this new feature is more flexible and easier to use.

> Why are we still far from the theoretical limit of ~600 milliseconds? Most plausible explanation is how occasional network hiccups are handled. A link may be congested for a quick moment, that would cause a packet or few to be lost, and the TCP stack would have to retransmit them once *RTO (retransmit timeout)* passes. This is usually barely noticeable in an already established connection, but there's often a huge difference when an initial SYN packet is lost. In this case, the TCP stack had no chance to estimate the optimal RTO yet, so it would have to start with a conservative default of *1 second*. Coming back to our workload, it's likely that while most of the clients were able to connect and subscribe in a fraction of a second, a couple of them tripped over SYN packet loss, which in turn caused the total time to increase by a second. Or, if we were extremely unlucky, two seconds.

Reviewer: ```suggestion
Why are we still far from the theoretical limit of ~600 milliseconds? The most plausible explanation is how occasional network hiccups are handled. A link may be congested for a quick moment, which would cause a packet or few to be lost, and the TCP stack would have to retransmit them once *RTO (retransmit timeout)* passes. This is usually barely noticeable in an already established connection, but there's often a huge difference when an initial SYN packet is lost. In this case, the TCP stack had no chance to estimate the optimal RTO yet, so it would have to start with a conservative default of *1 second*. Coming back to our workload, it's likely that while most of the clients were able to connect and subscribe in a fraction of a second, a couple of them tripped over SYN packet loss, which in turn caused the total time to increase by a second. Or, if we were extremely unlucky, two seconds.
```

If you find yourself deploying this kind of cluster for your geo-distributed application, definitely consider further [tuning the configuration](https://docs.emqx.com/en/emqx/v5.7/configuration/configuration.html). There are now a couple of extra configuration parameters worth tweaking to improve the performance of a multi-region EMQX cluster. Here is an `emqx.conf` example with parameters applicable to both core and replicant nodes combined for simplicity.

```shell
# Core nodes
node.default_bootstrap_batch_size = 10000
# Replicants
node.channel_cleanup_batch_size = 10
```

Now for the brief overview of what those parameters affect.

- `node.channel_cleanup_batch_size`

  Relevant for replicant nodes. If the network latency is high, significantly reducing this value from the default 10,000 improves performance during an abrupt disconnect of large numbers of clients.

- `node.default_bootstrap_batch_size`

  Relevant for core nodes. The default value is 500, but it can be greatly increased to reduce the time it takes for a replicant node to join a cluster with many active subscriptions.

## Alternatives

> Nevertheless, there are situations where latency this high is not acceptable. EMQX has a solution for these situations: rather than setting up a geo-distributed cluster, it is possible to deploy a separate independent cluster per each region and connect those clusters through [asynchronous MQTT bridges](https://docs.emqx.com/en/emqx/v5.7/data-integration/data-bridge-mqtt.html). It is a radically different deployment model, which demands additional compute resources and operational overhead. However, it offers a clear advantage: the shared state does not need to be consistently maintained across three continents. Each cluster will have its own state, and any changes induced by subscriptions will be handled locally, making the latency low.

Reviewer: ```suggestion
Nevertheless, there are situations where latency this high is not acceptable. EMQX has a solution for these situations: rather than setting up a geo-distributed cluster, it is possible to deploy a separate independent cluster per region and connect those clusters through [asynchronous MQTT bridges](https://docs.emqx.com/en/emqx/v5.7/data-integration/data-bridge-mqtt.html). It is a radically different deployment model, which demands additional compute resources and operational overhead. However, it offers a clear advantage: the shared state does not need to be consistently maintained across three continents. Each cluster will have its own state, and any changes induced by subscriptions will be handled locally, making the latency low.
```

Egress MQTT bridges in asynchronous mode are designed to interact with *external* resources, and in a way, remote EMQX clusters located on the other side of the globe are just that. Bridges have buffers backed by memory or durable storage, and they can handle intermittent connectivity issues often encountered in unreliable networks.

> Without a single global view of the connected clients and their subscriptions, it is no longer possible to route only a subset of messages to a particular region. Each MQTT bridge on each node will have to stream **whole** message flow to each remote location, saturating the egress bandwidth. It also inevitably comes with a loss of information: bridged messages will no longer have information about the original client. Besides, some effort is necessary to guarantee that the same bridged messages do not travel back and forth between the continents. However, [Rule engine](https://docs.emqx.com/en/emqx/v5.7/data-integration/rule-sql-syntax.html) should be expressive enough to handle this.

Reviewer: ```suggestion
Without a single global view of the connected clients and their subscriptions, it is no longer possible to route only a subset of messages to a particular region. Each MQTT bridge on each node will have to stream the **whole** message flow to each remote location, saturating the egress bandwidth. It also inevitably comes with a loss of information: bridged messages will no longer have information about the original client. Besides, some effort is necessary to guarantee that the same bridged messages do not travel back and forth between the continents. However, [Rule engine](https://docs.emqx.com/en/emqx/v5.7/data-integration/rule-sql-syntax.html) should be expressive enough to handle this.
```

> These shortcoming is what actually prompted us recently to work on another, more flexible solution: [Cluster Linking](https://github.com/emqx/emqx/pull/13126). The design goal was to combine the best of both worlds: have both the reliability and robustness of communication with external resources, and the ability to route only those messages a particular region is interested in, thus avoiding unnecessary waste of bandwidth and computing resources. This feature should see the light of day in the upcoming EMQX Enterprise 5.8.0 release.

Reviewer: ```suggestion
These shortcomings are what recently prompted us to work on another, more flexible solution: [Cluster Linking](https://github.com/emqx/emqx/pull/13126). The design goal was to combine the best of both worlds: have both the reliability and robustness of communication with external resources and the ability to route only those messages a particular region is interested in, thus avoiding unnecessary waste of bandwidth and computing resources. This feature should see the light of day in the upcoming EMQX Enterprise 5.8.0 release.
```
